### DISCUSSION TITLE
Usage without React Context with Next.js?

### DISCUSSION THREAD

**Discussion by dai-shi** [pmndrs/zustand#2740] [General]
Currently, our guide suggests to use React Context with Next.js.

This is to avoid misusing module state on the server (SSR).

https://github.com/pmndrs/zustand/blob/fec9135e33ec74469a8b2f8e543eb15a48a8ddc1/docs/guides/nextjs.md?plain=1#L28-L29

However, this only applies if we write to the store on the server. Is it correct?
In that case, if we can somehow prohibit updating the store on the server, is the usage of Zustand without React Context acceptable?

@dbritto-dev @charkour @TkDodo

**Comment by TkDodo** [comment]
I would say so, yes. The problem about creating a store on the server is that is will be shared between users. If we can't write to it, it will always be empty, so nothing would be shared.

**Reply by brianonchain** [reply]
Can you create a store on the client without using context?

**Comment by dbritto-dev** [comment]
@dai-shi I'd say that due to Next.js supports SSR (runtime) and SSG (build time) that's the safe way to create stores

**Reply by dai-shi** [reply]
Can you clarify a bit? What's the safe way to create stores?

**Reply by dbritto-dev** [reply]
@dai-shi Context API + `useContext` + `createStore`

**Reply by dai-shi** [reply]
Hm, so you do think it's unsafe with https://github.com/pmndrs/zustand/discussions/2740#discussioncomment-10649706 approach?

**Reply by dbritto-dev** [reply]
@dai-shi yeah

**Reply by dai-shi** [reply]
Can imagine a specific case that the approach breaks? Or, is it more like a hypothetical one?

**Reply by TkDodo** [reply]
@dai-shi I think the `initialState` for the store could still be derived from some request specific state? If you can only have a completely empty store, it doesn't really make sense to SSR that at all ...

**Reply by dai-shi** [reply]
But skipping SSR isn't very easy, right? Maybe there are several use cases. (Originally thought two but maybe more than that.)

**Reply by dai-shi** [reply]
> could still be derived from some request specific state?

Yeah, that's possible. It should be rare though.

Any idea to mitigate it? Just document so?

@dbritto-dev @TkDodo

**Reply by dbritto-dev** [reply]
@dai-shi we can document it

**Comment by dai-shi** [comment]
> if we can somehow prohibit updating the store on the server

One idea:

```js
import { create } from 'zustand';

const ssrSafe = (config, isSSR = typeof window === 'undefined') => (set, get, api) => {
  if (!isSSR) {
    return config(set, get, api);
  }
  const ssrSet = () => {
    throw new Error('Cannot set state of Zustand store in SSR');
  };
  api.setState = ssrSet;
  return config(ssrSet, get, api);
};

export const useFooStore = create(ssrSafe(...));
```

**Reply by dai-shi** [reply]
@dbritto-dev The discussion #2788 make me feel like pushing this idea.

**Reply by dai-shi** [reply]
See also #2326. cc @HasanMothaffar-OSG

**Reply by dai-shi** [reply]
PR #3271

**Reply by dai-shi** [reply]
#3308

**Comment by dai-shi** [comment]
#3055 suggests it should provide a way to hydrate.

So, here's one idea:

```js
const ssrSafe = config, isSSR = typeof window === 'undefined') => (set, get, api) => {
  if (!isSSR) {
    let hydrated = false;
    let initialState;
    const origGetInitialState = api.getInitialState;
    api.getInitialState = () => hydrated ? initialState : origGetInitialState();
    api.hydrate = (partial, replace) => {
      if (!hydrated) {
        api.setState(partial, replace);
        initialState = api.getState();
        hydrated = true;
      }
    };
    return config(set, get, api);
  }
  const ssrSet = () => {
    throw new Error('Cannot set state of Zustand store in SSR');
  };
  api.setState = ssrSet;
  return config(ssrSet, get, api);
};
```

**Comment by marcosm-dev** [comment]
Hi mates,
It says they’re going to update the guide for Next.js — does anyone know around what date?

Thanks

**Reply by dbritto-dev** [reply]
@marcosm-dev now we do have a new experimental middleware called `ssrSafe` -> https://github.com/pmndrs/zustand/blob/main/src/middleware/ssrSafe.ts

**Reply by marcosm-dev** [reply]
Thanks 👌🏻

**Reply by johncmunson** [reply]
Does this note still hold true?

```
// This is experimental middleware. It will be changed before finalizing it.
```

**Reply by dai-shi** [reply]
Yes. If you like you can create your own version on your end.

**Comment by sawa-ko** [comment]
Will the guide be updated someday, or will this no longer be done?

**Reply by dbritto-dev** [reply]
@sawa-ko working on that

**Reply by pauksztello** [reply]
ETA?

**Reply by dbritto-dev** [reply]
@pauksztello would be updated after migrating to rspress -> https://dbritto-dev.github.io/zustand/

**Reply by iamcrisb** [reply]
Hi @dbritto-dev I just wanted to install zustand in my nextjs 16.2 project and I saw that the docs are not up to date. What guide should I follow to properly configure zustand?

**Reply by dbritto-dev** [reply]
@iamcrisb here you go -> https://zustand.docs.pmnd.rs/learn/guides/nextjs

**Reply by iamcrisb** [reply]
@dbritto-dev it's saying that the doc should be updated based on this issue, my concern was that the issue is 18 months old.

**Reply by dbritto-dev** [reply]
@iamcrisb I don't get it, there's no issue, we don't recommend using it without context but you can use it if you make sure you don't write the store on server, that's it. Furthermore, context is needed when you want to isolated a store to a specific page/routes.

**Comment by DarkKingpro10** [comment]
I am not an expert but since the beginning I knew that Zustand can not be used in a RSC and Zustand was an alternative against React Context, but now you are saying that Zustand requires React Context, I marked the file as a client "use client" (The one that creates the store) and now it gives me error if I use it in a RSC.

<img width="979" height="434" alt="image" src="https://github.com/user-attachments/assets/3df772be-9fd7-435f-b622-d56110372df6" />

Because if I have to use Zustand with React Context, Why do I need Zustand in the first place?

Please teach me; I want to know if I am wrong

**Reply by DarkKingpro10** [reply]
I just remember the render problem of React Context, now I understand, but does my approach solve the problem, or is it not that way?

**Reply by dbritto-dev** [reply]
@DarkKingpro10 you can't use states on server components, at this point you should know that

**Reply by DarkKingpro10** [reply]
Dude, thats what I said,  I knew that Zustand can not be used in a RSC and Zustand was an alternative against React Context

I'm just saying that if Zustand is an alternative to React Context why should we use React Context to use Zustand.
Because if the problem is that someone will try to use Zustand in the server it will be a mistake, so what I did is mark the file that creates the store as a client component so It will show an error.

My question was, Did my approach solve the problem?
